cis194/week5/stevemao/Var.hs

```diff
@@ -0,0 +1,35 @@
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE FlexibleInstances #-}
+module Var where
+import qualified Data.Map as M
+import Homework
+
+class HasVars a where
+  var :: String -> a
+
+data VarExprT = Lit Integer
+           | Add VarExprT VarExprT
+           | Mul VarExprT VarExprT
+           | Var String
+  deriving (Show, Eq)
+
+instance Expr VarExprT where
+  lit = Lit
+  add = Add
+  mul = Mul
+
+instance HasVars VarExprT where
+  var = Var
+
+instance HasVars (M.Map String Integer -> Maybe Integer) where
+  var = M.lookup
+
+instance Expr (M.Map String Integer -> Maybe Integer) where
+  lit a = (\_ -> Just a)
+  add a b = (\m -> fmap (+) (a m) <*> (b m))
```

Comment:
```suggestion
  add a b = (\m -> (+) <$> (a m) <*> (b m)) -- this is a common pattern when using applicative. <$> followed by one or more <*>
```

Reply:
When to use applicative and when to use `fmap` and `chain`?

Reply:
When we can have `<$>` followed by one or `<*>`, we do `f <$> ... <*> ... <*> ...`.
`<$>` with `>>=` is not comparable to the above since they are with different type signature.
`f <$> a >>= g` need 2 functions `f` and `g`.

FYI, we call `>>=` `bind` in Haskell.